src/addons/metaservice/src/main/java/ee/ria/xroad/proxy/serverproxy/Openapi3Anonymiser.java

```diff
@@ -0,0 +1,105 @@
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import static ee.ria.xroad.common.ErrorCodes.X_INTERNAL_ERROR;
+
+/*
+ * Class for parsing Openapi3 description and replacing strings in servers.url-node with anonymised value
+ */
+public class Openapi3Anonymiser {
+
+    public static final String SERVERS = "servers";
+    public static final String URL = "url";
+    private static final ObjectMapper JSONMAPPER = new ObjectMapper(new JsonFactory());
+    private static final ObjectMapper YAMLMAPPER =
+            new ObjectMapper(new YAMLFactory()).configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true);
+
+    public void anonymiseJson(InputStream input, CachingStream output) throws IOException {
+        JsonNode tree = JSONMAPPER.readTree(input);
+        handleAnonymising(tree);
+        JSONMAPPER.writeValue(output, tree);
+    }
+
+    public void anonymiseYaml(InputStream input, CachingStream output) throws IOException {
+        JsonNode tree = YAMLMAPPER.readTree(input);
+        handleAnonymising(tree);
+        YAMLMAPPER.writeValue(output, tree);
+    }
+
+    /*
+     * Find servers.url section in the openapi3 specification.
+     * if the url does not contain variables then replace the url with the path part of the url
+     * if the url does contain variables then it is left as is
+     */
+    private void handleAnonymising(JsonNode tree) {
+        final JsonNode openapiVersion = tree.get("openapi");
+
+        // Check openapi version
+        if (openapiVersion != null && !openapiVersion.toString().startsWith("\"3.")) {
+            throw new CodedException(X_INTERNAL_ERROR,
+                    String.format("Incompatible openapi version. Openapi version 3 or greater expected. "
+                            + "Given openapi document is of version %s", openapiVersion.toString()));
+        }
+
+        final JsonNode servers = tree.get(SERVERS);
+        if (servers != null && servers.isArray()) {
+            servers.forEach(server -> {
+                if (server.has(URL) && !urlContainsVariables(server.get(URL))) {
+                    String url = server.get(URL).textValue();
+                    try {
+                        URI uri = new URI(url);
+                        ((ObjectNode) server).put(URL, uri.getPath());
+                    } catch (URISyntaxException e) {
+                        throw new CodedException(X_INTERNAL_ERROR,
+                                String.format("Can't parse url string: %s", url));
+                    }
+                }
+            });
+        }
+    }
+
+    // Return true if the url contains '{' and '}' meaning there's a variable in the url
+    private boolean urlContainsVariables(JsonNode url) {
+        String jsonString = url.toString();
```

Comment:
JsonNode.toString() renders the node (which can also be an object or array) as JSON.

Reply:
Changed parameter type to String